src/V3Param.cpp

```diff
@@ -52,14 +52,140 @@
 #include "V3Ast.h"
 #include "V3Case.h"
 #include "V3Const.h"
+#include "V3Os.h"
+#include "V3Parse.h"
 #include "V3Width.h"
 #include "V3Unroll.h"
 #include "V3Hashed.h"
 
 #include <deque>
 #include <map>
+#include <map>
 #include <vector>
 
+//######################################################################
+// Hierarchical block and parameter db (modules without parameter is also handled)
+class ParameterizedHierBlocks {
+    typedef std::multimap<string, const V3HierarchicalBlockOption*> HierBlockOptsByOrigName;
+    typedef HierBlockOptsByOrigName::const_iterator HierMapIt;
+    typedef std::map<string, AstNodeModule*> HierBlockModMap;
+    typedef std::map<string, AstConst*> ParamConstMap;
+    typedef std::map<const V3HierarchicalBlockOption*, ParamConstMap> ParamsMap;
+
+    // key:Original module name, value:HiearchyBlockOption*
+    // If a module is parameterized, the module is uniquiefied to overridden parameters.
+    // This is why HierBlockOptsByOrigName is multimap.
+    HierBlockOptsByOrigName m_hierBlockOptsByOrigName;
+
+    // key:mangled module name, value:AstNodeModule*
+    HierBlockModMap m_hierBlockMod;
+
+    // Overridden parameters of the hierarchical block
+    ParamsMap m_params;
+
+    static bool areSame(AstVar* modvarp, AstConst* varp, AstConst* paramp) {
```

Comment:
Seems general, perhaps make a AstConst isEpsilonEqual or something member function?   Suspect you don't need modvarp's type as can look at the const's dtypep() instead.

Reply:
Thanks for  the suggestion.
As you pointed out, `modvarp` was not necessary. I removed it in b249ff1784a639d7828f1f9fd0dc3bb29bc3b617.

I initially thought this function can be generalized, but then I remembered that the function has an assumption of
 - The first AstConst (varp) is before width analysis in V3Width, while the second AstConst (paramp) has exact width and type.

I'm afraid that it will be confusing if the function is a member of AstConst. (Casting and rounding when comparing int and real are not intuitive in general context.)

I added extra test data in 4ddbe6d296b8e78b63f74c50e08938fcd4d9d9a4 for those casting/rounding.